lib/valueflow.cpp

```diff
@@ -4415,28 +4470,78 @@ static void valueFlowContainerSize(TokenList *tokenlist, SymbolDatabase* symbold
 
             // possible value before condition
             valueFlowContainerReverse(scope.classDef, tok->varId(), value, settings);
+        }
+    }
+}
 
-            // possible value after condition
-            if (!isEscapeScope(scope.bodyStart, tokenlist, true)) {
-                const Token *after = scope.bodyEnd;
-                if (Token::simpleMatch(after, "} else {"))
-                    after = isEscapeScope(after->tokAt(2), tokenlist) ? nullptr : after->linkAt(2);
-                if (after && !isContainerSizeChanged(tok->varId(), scope.bodyStart, after))
-                    valueFlowContainerForward(after, tok->varId(), value, settings, tokenlist->isCPP());
-            }
+static void valueFlowContainerAfterCondition(TokenList *tokenlist,
+        SymbolDatabase *symboldatabase,
+        ErrorLogger *errorLogger,
+        const Settings *settings)
+{
+    ValueFlowConditionHandler handler;
+    handler.forward =
+    [&](Token *start, const Token *stop, const Variable *var, const std::list<ValueFlow::Value> &values, bool) {
+        // TODO: Forward multiple values
+        if (values.empty())
+            return false;
+        valueFlowContainerForward(start, var->declarationId(), values.front(), settings, tokenlist->isCPP());
+        return isContainerSizeChanged(var->declarationId(), start, stop);
+    };
+    handler.parse = [&](const Token *tok) {
+        ValueFlowConditionHandler::Condition cond;
+        ValueFlow::Value true_value;
+        ValueFlow::Value false_value;
+        const Token *vartok = parseCompareInt(tok, true_value, false_value);
+        if (vartok) {
+            vartok = vartok->tokAt(-3);
+            if(!isContainerSize(vartok))
+                return cond;
+            true_value.valueType = ValueFlow::Value::CONTAINER_SIZE;
+            false_value.valueType = ValueFlow::Value::CONTAINER_SIZE;
+            cond.true_values.push_back(true_value);
+            cond.false_values.push_back(false_value);
+            cond.vartok = vartok;
+            return cond;
+        }
 
-            // known value in conditional code
-            if (conditionToken->str() == "==" || conditionToken->str() == "(") {
-                const Token *parent = conditionToken->astParent();
-                while (parent && !Token::Match(parent, "!|==|!="))
-                    parent = parent->astParent();
-                if (!parent) {
-                    value.setKnown();
-                    valueFlowContainerForward(scope.bodyStart, tok->varId(), value, settings, tokenlist->isCPP());
-                }
+        // Empty check
+        if (tok->str() == "(") {
+            vartok = tok->tokAt(-3);
```

Comment:
this is a temporary hack until we want to handle structs/arrays/whatever. the day we want to handle that we need to use the ast. I am not against this as it is, I just wanted to mention it.